### DISCUSSION TITLE
Extend embedding module with Llama3 model

### DISCUSSION THREAD

**Discussion by tiadams** [SCAI-BIO/datastew#36] [Ideas]

**Comment by mehmetcanay** [comment]
Now it can be used from HugginFace with the model ID "meta-llama/Meta-Llama-3-8B-Instruct".

**Comment by tiadams** [comment]
> Now it can be used from HugginFace with the model ID "meta-llama/Meta-Llama-3-8B-Instruct".

We should test wether this performs better than MPNet (and also the time wise trade-off) and replace the MPNet default if it produces better results

**Comment by mehmetcanay** [comment]
Access to [meta-llama/Meta-Llama-3-8B-Instruct](https://huggingface.co/meta-llama/Meta-Llama-3-8B-Instruct) model is restricted. Also, it is no longer labeled as "sentence similarity" but "text generation". McGill has an alternative model [McGill-NLP/LLM2Vec-Meta-Llama-3-8B-Instruct-mntp-supervised](https://huggingface.co/McGill-NLP/LLM2Vec-Meta-Llama-3-8B-Instruct-mntp-supervised), but we need to write a new adapter as this model does not utilize SentenceTransformer module.

**Comment by tiadams** [comment]
We need in general a well performing model that can map from other languages (e.g. german) to the english terminologies we are using.

Doesn't need to be llama specifically but we need some sort of alternative to MPNet since it can only handle english to english.

**Comment by mehmetcanay** [comment]
I am looking into it. For now, we can test [FremyCompany/BioLORD-2023-M](https://huggingface.co/FremyCompany/BioLORD-2023-M) as it can handle semantic similarity in a multi-lingual context. This also does not require a new adapter.

**Comment by mehmetcanay** [comment]
[FremyCompany/BioLORD-2023-M](https://huggingface.co/FremyCompany/BioLORD-2023-M) could not handle our CDM for an unknown reason. For other models I have found that work with our adapter I re-ran the workflow I had for translated BIOFIND dictionary (145 variables). I put the old models we have tested for comparison. Here are the results:

|         | Enligh to German | German to English | Average |
|---------|---------------|----------------|---------|
| [text-embedding-3-large model](https://openai.com/index/new-embedding-models-and-api-updates/) | 0.46          | 0.53           | 0.50    |
| [sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2](https://huggingface.co/sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2)  | 0.19          | 0.28           | 0.24    |
| [sentence-transformers/distiluse-base-multilingual-cased-v1](https://huggingface.co/sentence-transformers/distiluse-base-multilingual-cased-v1)      | 0.23          | 0.23           | 0.23    |
| [sentence-transformers/distiluse-base-multilingual-cased-v2](https://huggingface.co/sentence-transformers/distiluse-base-multilingual-cased-v2)      | 0.15          | 0.25           | 0.20    |
| [sentence-transformers/all-mpnet-base-v2](https://huggingface.co/sentence-transformers/all-mpnet-base-v2)   | 0.12          | 0.19           | 0.16    |
| [FremyCompany/BioLORD-2023](https://huggingface.co/FremyCompany/BioLORD-2023) | 0.14          | 0.13           | 0.14    |

Likewise, I tested the new models with our harmonization workflow:

|         |Average |
|---------|---------|
| [text-embedding-3-large model](https://openai.com/index/new-embedding-models-and-api-updates/)  | 0.77    |
| [FremyCompany/BioLORD-2023](https://huggingface.co/FremyCompany/BioLORD-2023) |0.76    |
| [sentence-transformers/all-mpnet-base-v2](https://huggingface.co/sentence-transformers/all-mpnet-base-v2)   | 0.73    |
| [sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2](https://huggingface.co/sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2)  | 0.72    |
| [sentence-transformers/distiluse-base-multilingual-cased-v1](https://huggingface.co/sentence-transformers/distiluse-base-multilingual-cased-v1)      | 0.61    |
| [sentence-transformers/distiluse-base-multilingual-cased-v2](https://huggingface.co/sentence-transformers/distiluse-base-multilingual-cased-v2)      | 0.60    |

**Comment by tiadams** [comment]
> [FremyCompany/BioLORD-2023-M](https://huggingface.co/FremyCompany/BioLORD-2023-M) could not handle our CDM for an unknown reason. For other models I have found that work with our adapter I re-ran the workflow I had for translated BIOFIND dictionary (145 variables). I put the old models we have tested for comparison. Here are the results:
> 	Enligh to German 	German to English 	Average
> [text-embedding-3-large model](https://openai.com/index/new-embedding-models-and-api-updates/) 	0.46 	0.53 	0.50
> [sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2](https://huggingface.co/sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2) 	0.19 	0.28 	0.24
> [sentence-transformers/distiluse-base-multilingual-cased-v1](https://huggingface.co/sentence-transformers/distiluse-base-multilingual-cased-v1) 	0.23 	0.23 	0.23
> [sentence-transformers/distiluse-base-multilingual-cased-v2](https://huggingface.co/sentence-transformers/distiluse-base-multilingual-cased-v2) 	0.15 	0.25 	0.20
> [sentence-transformers/all-mpnet-base-v2](https://huggingface.co/sentence-transformers/all-mpnet-base-v2) 	0.12 	0.19 	0.16
> [FremyCompany/BioLORD-2023](https://huggingface.co/FremyCompany/BioLORD-2023) 	0.14 	0.13 	0.14
> 
> Likewise, I tested the new models with our harmonization workflow:
> 	Average
> [text-embedding-3-large model](https://openai.com/index/new-embedding-models-and-api-updates/) 	0.77
> [FremyCompany/BioLORD-2023](https://huggingface.co/FremyCompany/BioLORD-2023) 	0.76
> [sentence-transformers/all-mpnet-base-v2](https://huggingface.co/sentence-transformers/all-mpnet-base-v2) 	0.73
> [sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2](https://huggingface.co/sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2) 	0.72
> [sentence-transformers/distiluse-base-multilingual-cased-v1](https://huggingface.co/sentence-transformers/distiluse-base-multilingual-cased-v1) 	0.61
> [sentence-transformers/distiluse-base-multilingual-cased-v2](https://huggingface.co/sentence-transformers/distiluse-base-multilingual-cased-v2) 	0.60

Thanks for the research, great work. I think based in this we should probably switch to the [text-embedding-3-large model](https://openai.com/index/new-embedding-models-and-api-updates/) since it appears to perform best in both metrics

**Comment by tiadams** [comment]
Just saw this is openAI based, probably best trade off then is [sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2](https://huggingface.co/sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2), what do you think?

**Reply by mehmetcanay** [reply]
This makes sense to me as well, it is a great trade-off between overall accuracy and multilingual capability compared to [sentence-transformers/all-mpnet-base-v2](https://huggingface.co/sentence-transformers/all-mpnet-base-v2).

**Comment by tiadams** [comment]
@mehmetcanay 

In the last CePPH project meeting someone proposed this one for german/english mapping:
https://huggingface.co/nomic-ai/nomic-embed-text-v1.5

Have you seen this one yet? It is also available as part of sentence-transformers library so we could also use this

**Reply by mehmetcanay** [reply]
Cool! I didn't see this, I will try it and extend the table we have.

**Reply by mehmetcanay** [reply]
This requires several modifications to our class and the input. 

- We need to add the `trust_remote_code=True` argument to the initialization of the `SentenceTransformer`class.
- We need to append a task instruction prefix (e.g., `"search_query: "`) to the beginning of each sentence passed to the model.
- We need to add `einops` to the requirements

**Comment by tiadams** [comment]
@mehmetcanay seen this:
https://github.com/SeanLee97/AnglE?

Looks promising, we could use this as a wrapper to solve #21

**Reply by mehmetcanay** [reply]
I didn't see this before, looks very cool and will definitely allow us to simplify. But majority of the models we have tested are not listed in the README of the project, are we sure that it will support the models we want?

**Comment by tiadams** [comment]
Langchains API also looks like a good option:
https://python.langchain.com/docs/integrations/text_embedding/

**Reply by mehmetcanay** [reply]
This looks like a better option as it already supports Huggin Face models and Open AI. Plus the Llama3 model is also implemented and can be used free of charge.

**Reply by mehmetcanay** [reply]
Will we be able to generate embeddings in batches with this API in-between though, or is it going to be a trade-off between simplicity and time required for calculation?

**Reply by mehmetcanay** [reply]
I just checked the [article](https://www.nature.com/articles/s41746-024-01344-w) Martin sent us. Apparently MistralAI shares their weights publicly which makes them semi-open source. This could be interesting to the user.

**Reply by mehmetcanay** [reply]
Also OLMo 7B Instruct model regarded as the most open-source model in the same article. The model exists in Hugging Face, but I am not sure whether it can be utilized for Sentence Similarity without doing anything additional.

Edit: I am adding the figure as well so we can also see the criteria.

![image](https://github.com/user-attachments/assets/6c66083f-3026-4cad-9498-20f8c8cce7f3)
